src/mame/nintendo/cham24.cpp

```diff
@@ -114,61 +113,45 @@ class cham24_state : public driver_device
 
 void cham24_state::cham24_set_mirroring(int mirroring)
 {
-	switch (mirroring)
-	{
-		case PPU_MIRROR_HORZ:
-			for (int i = 0; i < 4; i++)
-				m_nt_page[i]->set_entry(BIT(i, 1));
-			break;
-		case PPU_MIRROR_VERT:
-		default:
-			for (int i = 0; i < 4; i++)
-				m_nt_page[i]->set_entry(i & 1);
-			break;
-	}
-}
+	int bit = mirroring == PPU_MIRROR_HORZ;
 
-void cham24_state::sprite_dma_w(address_space &space, uint8_t data)
-{
-	int source = (data & 7);
-	m_ppu->spriteram_dma(space, source);
+	for (int i = 0; i < 4; i++)
+		m_nt_page[i]->set_entry(BIT(i, bit));
 }
 
-uint8_t cham24_state::cham24_IN0_r()
+template <u8 Which>
+u8 cham24_state::cham24_in_r()
 {
-	return ((m_in_0 >> m_in_0_shift++) & 0x01) | 0x40;
+	if (m_input_strobe)
+		m_input_latch[Which] = m_in[Which]->read();
+
+	u8 ret = 0x40;
+
+	ret |= m_input_latch[Which] & 1;
+	m_input_latch[Which] >>= 1;
```

Comment:
Is this a side effect of a read that isn’t properly suppressed?

Reply:
Not sure I understand the question. It's a direct effect of the output enable pin clocking the shift register. At least that's what it is on official hardware, so I've mimicked in this bootleg having no other info to go on.

Reply:
It's a read handler worth side effects. You generally need to suppress side effects of reads for the debugger unless you want memory views to wreak havoc.